iree/base/internal/dynamic_library_win32.c

```diff
@@ -0,0 +1,425 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <stdio.h>
+
+#include "iree/base/internal/atomics.h"
+#include "iree/base/internal/dynamic_library.h"
+#include "iree/base/internal/file_path.h"
+#include "iree/base/target_platform.h"
+#include "iree/base/threading.h"
+#include "iree/base/tracing.h"
+
+#if defined(IREE_PLATFORM_WINDOWS)
+
+// TODO(benvanik): support PDB overlays when tracy is not enabled; we'll
+// need to rearrange how the dbghelp lock is handled for that (probably moving
+// it here and having the tracy code redirect to this).
+#if defined(TRACY_ENABLE)
+#define IREE_HAVE_DYNAMIC_LIBRARY_PDB_SUPPORT 1
+#pragma warning(disable : 4091)
+#include <dbghelp.h>
+extern "C" void IREEDbgHelpLock();
+extern "C" void IREEDbgHelpUnlock();
+#endif  // TRACY_ENABLE
+
+struct iree_dynamic_library_s {
+  iree_atomic_ref_count_t ref_count;
+  iree_allocator_t allocator;
+
+  // Base module name used as an identifier. When loaded from a file this must
+  // be the basename for dbghelp to be able to find symbols.
+  // Owned and allocated as part of the struct upon creation.
+  // Has NUL terminator for compatibility with Windows APIs.
+  char* identifier;
+
+  // File path of the loaded module, if loaded from one.
+  // Owned and allocated as part of the struct upon creation.
+  // Has NUL terminator for compatibility with Windows APIs.
+  char* module_path;
+
+  // Windows module handle.
+  HMODULE module;
+
+  // 0 or more file paths that were created as part of the loading of the
+  // library or attaching of symbols from memory.
+  //
+  // Each path string is allocated using the |allocator| and freed during
+  // library deletion.
+  iree_host_size_t temp_file_count;
+  char* temp_file_paths[2];
+};
+
+static iree_once_flag iree_dynamic_library_temp_path_flag_ =
+    IREE_ONCE_FLAG_INIT;
+static char iree_dynamic_library_temp_path_base_[MAX_PATH + 1];
+static void iree_dynamic_library_init_temp_paths(void) {
+  // Query the temp path from the OS. This can be overridden with the following
+  // environment variables: [TMP, TEMP, USERPROFILE].
```

Comment:
Same comment here re: `TEST_TMPDIR`

https://github.com/google/iree/blob/603b208f623e65eb48d8594c4ffc4218f36f41f5/iree/base/internal/file_io_win32.cc#L104-L108

(Less important here since we don't have too many developers or build bots running tests exercising this :P)